src/inmanta/server/services/orchestrationservice.py

```diff
@@ -419,6 +427,158 @@ def safe_get(input: JsonType, key: str, default: object) -> object:
 
         return 200
 
+    @handle(methods.put_version, env="tid")
+    async def put_version(
+        self,
+        env: data.Environment,
+        version: int,
+        resources: List[JsonType],
+        resource_state: Dict[ResourceIdStr, const.ResourceState],
+        unknowns: List[Dict[str, PrimitiveTypes]],
+        version_info: JsonType,
+        compiler_version: Optional[str] = None,
+        resource_sets: Optional[Dict[ResourceIdStr, Optional[str]]] = None,
+    ) -> Apireturn:
+        return await self._put_version(
+            env, version, resources, resource_state, unknowns, version_info, compiler_version, resource_sets
+        )
+
+    @handle(methods_v2.put_partial, env="tid")
+    async def put_partial(
+        self,
+        env: data.Environment,
+        version: int,
+        resources: list[Any],
```

Comment:
If I understand correctly this contains resources which we cannot really type correctly because we would need support for JsonType (Dict[str, Any]). So there are two options: allowing this in protocol or using kwargs.

In the case of kwargs you indeed need to validate everything yourself. The way I did that was creating a pydantic object for all the arguments and creating that (with the option that additional attributes are not allowed)

This would indeed mean that you add `**kwargs: object` also to the handler and than do `a = custom_pydantic_class(**kwargs); a.resources`

Reply:
But when the field names are known ahead of time, why would we use `**kwargs` rather than `myfield: object`? You still need to validate it yourself but we at least still have a simple signature.

Reply:
If object works I would use that, but iirc protocol also does not allow that.

Reply:
I see. I think it would be beneficial to add support (@FloLey not in this PR, don't worry). It's not perfect because it circumvents type checks but it does so in a similar manner as `**kwargs` while it has some advantages over the latter.

What about the second part of my comment? As I understood it protocol should check the type signature of kwargs fields in the method handlers and raise an exception if they're not typed as `object`. Did I misunderstand that, or is this check just not working as intended?

Reply:
Protocol validates what does in, it does not check what the handlers itself define. There is "trust" that the handlers are correct.